holoviews/tests/operation/test_operation.py

```diff
@@ -126,13 +232,71 @@ def test_qmesh_curvilinear_edges_contours(self):
+                 (0.25, -0.25, 1), (0.25, 0.25, 1), (0.08333333, 0.25, 1), (0.08333333, -0.25, 1)]]
+        polys = Polygons(data, vdims=img.vdims[0].clone(range=(0.5, 1.5)))
+        self.assertEqual(op_contours, polys)
+
+    def test_image_contours_filled_with_hole(self):
+        img = Image(np.array([[0, 0, 0], [0, 1, 0.], [0, 0, 0]]))
+        # Single polygon with hole
+        op_contours = contours(img, filled=True, levels=[0.25, 0.75])
+        data = [[(-0.25, 0.0, 0.5), (0.0, -0.25, 0.5), (0.25, 0.0, 0.5), (0.0, 0.25, 0.5),
+                  (-0.25, 0.0, 0.5)]]
+        polys = Polygons(data, vdims=img.vdims[0].clone(range=(0.25, 0.75)))
+        self.assertEqual(op_contours, polys)
+        expected_holes = [[[np.array([[0.0, -0.08333333], [-0.08333333, 0.0], [0.0,  0.08333333],
+                                      [0.08333333, 0.0], [0.0, -0.08333333]])]]]
+        np.testing.assert_array_almost_equal(op_contours.holes(), expected_holes)
+
+    def test_image_contours_filled_multi_holes(self):
+        img = Image(np.array([[0, 0, 0, 0, 0], [0, 1, 0, 1, 0], [0, 0, 0, 0, 0]]))
+        # Single polygon with two holes
+        op_contours = contours(img, filled=True, levels=[-0.5, 0.5])
+        data = [[(-0.4, -0.3333333, 0), (-0.2, -0.3333333, 0), (0, -0.3333333, 0),
+                 (0.2, -0.3333333, 0), (0.4, -0.3333333, 0), (0.4, 0, 0), (0.4, 0.3333333, 0),
+                 (0.2, 0.3333333, 0), (0, 0.3333333, 0), (-0.2, 0.3333333, 0), (-0.4, 0.3333333, 0),
+                 (-0.4, 0, 0), (-0.4, -0.3333333, 0)]]
+        polys = Polygons(data, vdims=img.vdims[0].clone(range=(-0.5, 0.5)))
+        self.assertEqual(op_contours, polys)
+        expected_holes = [[[np.array([[-0.2, -0.16666667], [-0.3, 0], [-0.2, 0.16666667], [-0.1, 0],
+                                      [-0.2, -0.16666667]]),
+                            np.array([[0.2, -0.16666667], [0.1, 0], [0.2, 0.16666667], [0.3, 0],
+                                      [0.2, -0.16666667]])]]]
+        np.testing.assert_array_almost_equal(op_contours.holes(), expected_holes)
+
+    def test_image_contours_filled_empty(self):
         img = Image(np.array([[0, 1, 0], [3, 4, 5.], [6, 7, 8]]))
-        op_contours = contours(img, filled=True, levels=[2, 2.5])
-        data = [[(0., 0.166667, 2.25), (0.333333, 0.166667, 2.25), (0.333333, 0.2, 2.25), (0., 0.222222, 2.25),
-                 (-0.333333, 0.111111, 2.25), (-0.333333, 0.055556, 2.25), (0., 0.166667, 2.25)]]
-        polys = Polygons(data, vdims=img.vdims[0].clone(range=(2, 2.5)))
+        # Contour level outside of data limits
+        op_contours = contours(img, filled=True, levels=[20.0, 23.0])
+        polys = Polygons([[]], vdims=img.vdims[0].clone(range=(20.0, 23.0)))
         self.assertEqual(op_contours, polys)
 
+    def test_image_contours_filled_auto_levels(self):
+        z = np.array([[0, 1, 0], [3, 4, 5.], [6, 7, 8]])
```

Comment:
Is the `5.` to check if I review thoroughly enough? :upside_down_face:

Reply:
That is copied and pasted from previous image tests. I have confirmed that it gives the same results with or without the decimal point.